ansible/cloud_providers/gcp_infrastructure_deployment.yml

```diff
@@ -0,0 +1,193 @@
+---
+- name: Step 001 Deploy Infrastructure
+  hosts: localhost
+  connection: local
+  gather_facts: false
+  become: false
+  tags:
+    - step001
+    - deploy_infrastructure
+  tasks:
+    - name: Ensure gcloud is installed
+      environment:
+        PATH: /usr/bin
+      command: which gcloud
+      register: gcloud_result
+
+    - name: Fail if gcloud not available
+      fail:
+        msg: you need Google Cloud SDK installed
+      when: gcloud_result is failed
+
+    - name: Get SSH public key
+      set_fact:
+        ssh_key: "~/.ssh/{{key_name}}.pem"
+        ssh_key_data: "{{lookup('file', '~/.ssh/{{key_name}}.pub')}}"
```

Comment:
Do we have to use a local key ?

We do that for AWS because it's not possible to create a key along with a cloudformation stack.

But for OSP for example, we create an infra ssh key inside the HEAT template and use it for the deployment.

In other words, is it possible to create the infra key inside the gcp template ?
if yes, then i'd like to get rid of any stateful code like this one

```yaml
ssh_key: "~/.ssh/{{key_name}}.pem"
```

And use the key from the template.

Logic is:

- create the stack
- get the key from the output of the stack
- save it to `output_dir`
- use that key instead, and not something from `~/.ssh/`

Reply:
Templates aren't that smart in Google, but I will look at the code in OSP to see if I can find a middle ground.

Reply:
I see how you are doing it with your heat templates and I'm not sure how'd I'd make that work with Google.
Google does create a private set of ssh keys which can be used to access the hosts but you need to connect via "gcloud compute ssh -z zone1b instance1" instead of just ssh from the command line. That would be a lot of modification to a lot of roles when we can just use this pattern which matches how Azure does it.
Luckily Alibaba works like AWS with stored keys. so I won't need to use this pattern for the next cloud I will likely enable.

Reply:
Can we instead of having the key as a requirement, generate it for each deployment ?

IMO, the "infra" key for the environment should be per-deployment.

> Google does create a private set of ssh keys which can be used to access the hosts but you need to connect via "gcloud compute ssh -z zone1b instance1" instead of just ssh from the command line
I see.
We can generate the key in `output_dir` using basic `ssh-keygen` commands, see https://github.com/redhat-cop/agnosticd/blob/development/ansible/roles/infra-local-create-ssh_key/tasks/main.yml  

Regarding extra access to the environment, i would prefer if we just use regular authorized_keys like this:

https://github.com/redhat-cop/agnosticd/blob/85463267058cc276631aacbffb0b89935cae0427/ansible/roles/infra-osp-template-generate/defaults/main.yaml#L4-L6


That way:
1. we have an infra ssh key generated per-deployment
2. if we need to, we can add more access using the pub keys of admin

wdyt ?

Reply:
We can add keys to the authorized_keys file, that's not a problem.

And I will build some logic towards having the infra_key always generated and also apply it to Azure over the next couple days. It can be done in agnosticD, just not in the Google template like you do with Heat.

Reply:
its a user:key parameter in the template file, so we can make a user account for for every admin key or add all the keys to a single user like opentlc-mgr.

Reply:
I have the admin keys all being added to the {{ remote_user }} authorized_keys file for Azure and GCP along with an auto generated ssh_key

Reply:
Added the public ssh keys to default_vars.yml so you can easily override it